model-optimizer/extensions/ops/If.py

```diff
@@ -0,0 +1,321 @@
+        for node in subgraph.get_op_nodes():
+            if node.has('input_id'):
+                assert node.soft_get('type') == 'Parameter'
+                input_port_id = node['input_id']
+                input_type = if_node.in_port(input_port_id).get_data_type()
+                node.data_type = input_type
+                log.debug('Updated data type for the body node with name "{}" with value {}'
+                          .format(node.name, node.data_type))
+
+    @staticmethod
+    def update_body_parameters_shape(if_node: Node, condition: bool):
+        """
+        Update shape for If body parameters.
+
+        :param if_node: The If node
+        :param condition: the boolean defining a condition (then/else) graph to add connect the body
+        :return: None
+        """
+        subgraph = if_node.then_graph if condition else if_node.else_graph
+        for node in subgraph.get_op_nodes():
+            if node.has('input_id'):
+                assert node.soft_get('type') == 'Parameter'
+                input_port_id = node['input_id']
+                input_shape = if_node.in_port(input_port_id).data.get_shape()
+                if node.soft_get('shape', None) is None:
+                    node['shape'] = None
+                node.shape = input_shape.copy()
+                log.debug('Updated shape for the body node with name "{}" with value {}'
+                          .format(node.soft_get('name', node.soft_get('id')), node.shape))
+
+    @staticmethod
+    def update_if_output_ports_shape(if_node: Node):
+        """
+        Update shape and values for If output ports.
+
+        :param if_node: The If node to update output ports and shapes
+        :return: None
+        """
+
+        then_outputs = [node for node in if_node.then_graph.get_op_nodes() if node.has('output_id')]
+        else_outputs = [node for node in if_node.else_graph.get_op_nodes() if node.has('output_id')]
+        outputs_mapping = {}
+        outputs_number = len(if_node.out_ports())
+
+        if outputs_number == 0 and len(if_node.out_ports(control_=True)) != 0:
+            # Some models have if with control flow outputs.
+            # These shape inference for such ifs
+            # TODO: need to rethink and redo support for control flow edges in if operation
+            for node in if_node.out_nodes(True).values():
+                node.shape = int64_array([])
+            return
+
+        for port_id in if_node.out_ports().keys():
+            outputs_mapping[port_id] = {}
+        port_ids = outputs_mapping.keys()
+
+        # variable then_contains_fake_outputs/else_contains_fake_outputs contains True value
+        # if all outputs from then_body/else_body have shape [0]. It means then_body/else_body does not return data
+        # and further shape_inference for this branch is not possible.
+        # TODO: exclude support fake_outputs from this code when we will support shape_inference with empty tensors
```

Comment:
these are not about fake outputs then. I would rather call it as `is_then_fake` and `is_else_fake` (or `_dummy`).

Reply:
If we support shape_infer for each op in mo for empty tensors we will not need fake outputs. Not it is neccessary because our operations cannot work with empty tensors.

Reply:
even if we get all outputs as empty tensors, shape inference will be work correct in future